rfd/0000-sre-challenge.md

```diff
@@ -0,0 +1,213 @@
+### Design Document for Kubernetes Deployment Scaler
+
+---
+authors: Derek Norrbom (dereknorrbom@gmail.com)
+state: draft
+---
+
+## What
+
+This document outlines the proposed design and implementation plan for a Kubernetes Deployment Scaler API. The API will provide functionality to get and set the replica count of a Kubernetes deployment, utilizing mTLS for secure communication and caching for efficient data retrieval.
+
+## Why
+
+The purpose of this project is to implement a scalable, secure API for managing Kubernetes deployments. This project will demonstrate proficiency in Go, Kubernetes, Docker, and security best practices (mTLS). It will also highlight the ability to design, build, and document an application that can be easily extended and maintained.
+
+## How
+
+### API Structure
+
+The API will provide the following endpoints:
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running.
+- **GET /replica-count**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count**: Sets the replica count of the specified deployment.
+- **GET /deployments**: Lists all deployments in the specified namespace.
+
+Each endpoint will be secured using mutual TLS (mTLS) to ensure secure communication between the client and the server.
+
+### Developer Workflow
+
+1. **Clone the Repository**: 
+   ```
+   git clone <repository-url>
+   cd k8s-deployment-scaler
+   ```
+
+2. **Run the Setup Script**:
+   A `setup.sh` script will be provided to automate the detection and installation of the required dependencies (Docker, KIND, Helm, kubectl (Kubernetes command-line tool)).
```

Comment:
What kind of OS will your script target?

Installation for those tools might be slightly different for MacOS/Linux/Windows

Reply:
The setup.sh script will target x86 64-bit Linux and MacOS environments, as specified in the challenge documentation.  Windows is out of scope for this project, so the script will not include Windows-specific installation instructions.